### PR TITLE
Support custom block and filesystem export for all drivers

### LINKED ISSUE #9064: Support custom block and filesystem export for all drivers
It looks like when custom volumes gained block content type support (https://github.com/lxc/lxd/pull/7479) that no provision was made to update the storage drivers to support exporting those volume types.

As such doing something like:

```
lxc storage volume create lvm vol1 --type=block
lxc storage volume export lvm vol1 /home/user/vol1.tar.gz
```

Results in a basically empty tarball with no block image files present. 
Also affects ZFS and likely any block backed storage driver (such as LVM, ZFS, and Ceph).

BTRFS and Dir drivers seem to work, but probably by fluke (because they already have block image files on disk) rather than by design.

So they should be tested for proper operation nonetheless. 

- [ ] Fix block volume exports for all drivers
- [ ] Add support for filesystem exports for cephfs driver (see https://discuss.linuxcontainers.org/t/lxd-cephfs-volume-export-results-in-error-not-implemented/11906)

### PR DESCRIPTION
**Pull request by monstermunchkin**
This resolves #9064.

### CODE REVIEW

**Review by tomponline — COMMENTED** [review]
I'll go through this in a bit, but can you add an update to the `lxc-ci` tests for these changes to avoid any regressions in the future. Thanks

**Comment by monstermunchkin** [comment]
> And for filesystem custom volumes we could just update the internal test suite in lxd.

What exactly do we need to update here? We have tests for storage volume export/import in `backup.sh`.

**Comment by tomponline** [comment]
> > And for filesystem custom volumes we could just update the internal test suite in lxd.
> 
> What exactly do we need to update here? We have tests for storage volume export/import in `backup.sh`.

the way the PR title is worded suggests that something has been fixed/added for filesystem volumes too (that implies it wasn't previously tested because the tests aren't passing :)).

**Comment by tomponline** [comment]
Also the original issue https://github.com/lxc/lxd/issues/9064 mentions adding filesystem export for cephfs custom volumes. I seem to remember that got added, but wonder if its viable to add tests for that to the lxd test suite (if not already)?

**Comment by monstermunchkin** [comment]
> the way the PR title is worded suggests that something has been fixed/added for filesystem volumes too (that implies it wasn't previously tested because the tests are passing :)).

I just copied the title from the actual issue :wink:

**Comment by tomponline** [comment]
> > the way the PR title is worded suggests that something has been fixed/added for filesystem volumes too (that implies it wasn't previously tested because the tests are passing :)).
> 
> I just copied the title from the actual issue wink

Indeed, so adding tests for cephfs custom volume export is needed to be sure it works for all filesystems before closing #9064

File: lxd/storage/drivers/generic_vfs.go [hunk 1]
```diff
@@ -496,37 +492,36 @@ func genericVFSBackupVolume(d Driver, vol Volume, tarWriter *instancewriter.Inst
...
-					return nil
-				})
-				if err != nil {
-					return err
 				}
 
 				name := fmt.Sprintf("%s.%s", prefix, genericVolumeBlockExtension)
 
-				logMsg = "Copying virtual machine block volume"
+				logMsg := "Copying virtual machine block volume"
```
Comment (tomponline) [inline lxd/storage/drivers/generic_vfs.go:524]: this message needs to change based on whether the volume is VMBlock or not, similar to below:

```go
			logMsg := "Unpacking virtual machine block volume"
			if vol.volType == VolumeTypeCustom {
				logMsg = "Unpacking custom block volume"
			}
```

**Review by tomponline — CHANGES_REQUESTED** [review]

**Comment by tomponline** [comment]
Other than small logging issue, this all looks good! :)

**Review by tomponline — CHANGES_REQUESTED** [review]
Also the cephfs tests are failing:

```
lxc storage volume import lxdtest-goF-cephfs /lxc-ci/build/tmp.mlpP6XtO01/go/src/github.com/lxc/lxd/test/tmp.9pw/goF/testvol.tar.gz
+ LXC_LOCAL=1 lxc_remote storage volume import lxdtest-goF-cephfs /lxc-ci/build/tmp.mlpP6XtO01/go/src/github.com/lxc/lxd/test/tmp.9pw/goF/testvol.tar.gz
+ set +x
+ eval /lxc-ci/build/tmp.mlpP6XtO01/go/bin/lxc "storage" "volume" "import" "lxdtest-goF-cephfs" "/lxc-ci/build/tmp.mlpP6XtO01/go/src/github.com/lxc/lxd/test/tmp.9pw/goF/testvol.tar.gz" --verbose
+ /lxc-ci/build/tmp.mlpP6XtO01/go/bin/lxc storage volume import lxdtest-goF-cephfs /lxc-ci/build/tmp.mlpP6XtO01/go/src/github.com/lxc/lxd/test/tmp.9pw/goF/testvol.tar.gz --verbose
Error: Create custom volume from backup: Error starting unpack: Failed to run: tar -zxf - --xattrs-include=* -C /lxc-ci/build/tmp.mlpP6XtO01/go/src/github.com/lxc/lxd/test/tmp.9pw/goF/storage-pools/lxdtest-goF-cephfs/custom/default_testvol --transform=s/^backup\/volume-snapshots\/snap0/./ backup/volume-snapshots/snap0: tar: .: Cannot open: File exists
tar: Exiting with failure status due to previous errors
```

File: lxd/storage/drivers/generic_vfs.go (hunk 1 above)
Comment (monstermunchkin) [reply lxd/storage/drivers/generic_vfs.go:524]: It already does.

Comment (tomponline) [reply lxd/storage/drivers/generic_vfs.go:524]: So it does, dont know why I didn't see that.

**Comment by monstermunchkin** [comment]
I found the issue with cephfs volume backup.There's a single line we'd need to add to `genericVFSBackupVolume()` to make it work. This line would is: `snapVol.mountCustomPath = filepath.Join(vol.MountPath(), ".snap", snapName)`.

Now, since `generic_vfs.go` is generic, we want to avoid checks like `if d.name == "cephfs"`, right?

An alternative would be a new driver function `VolumeSnapshotMountPath()` which returns the real mount path, which could be used in `genericVFSBackupVolume()`.

@tomponline What are your thoughts on this?

**Review by tomponline — APPROVED** [review]

**Comment by tomponline** [comment]
For reference, we have decided to tackle the cephfs issue in a separate PR.

**Comment by stgraber** [comment]
jenkins: test this please

**Pull request merged**